src/check/canonicalize.zig

```diff
@@ -85,6 +104,51 @@ pub fn get_gpa(self: *Self) std.mem.Allocator {
     comptime return self.can_ir.env.gpa;
 }
 
+/// Enter a function boundary by pushing its region onto the stack
+pub fn enterFunction(self: *Self, region: Region) void {
+    self.function_regions.append(self.can_ir.env.gpa, region) catch |err| exitOnOom(err);
+}
+
+/// Exit a function boundary by popping from the stack
+pub fn exitFunction(self: *Self) void {
+    _ = self.function_regions.pop();
+}
+
+/// Get the current function region (the function we're currently in)
+pub fn getCurrentFunctionRegion(self: *const Self) ?Region {
+    if (self.function_regions.items.len > 0) {
+        return self.function_regions.items[self.function_regions.items.len - 1];
+    }
+    return null;
+}
+
+/// Record which function a var pattern was declared in
+pub fn recordVarFunction(self: *Self, pattern_idx: CIR.Pattern.Idx) void {
+    // Mark this pattern as a var
+    self.var_patterns.put(self.can_ir.env.gpa, pattern_idx, {}) catch |err| exitOnOom(err);
+
+    if (self.getCurrentFunctionRegion()) |function_region| {
+        self.var_function_regions.put(self.can_ir.env.gpa, pattern_idx, function_region) catch |err| exitOnOom(err);
+    }
+}
+
+/// Check if a pattern is a var
+pub fn isVarPattern(self: *const Self, pattern_idx: CIR.Pattern.Idx) bool {
+    return self.var_patterns.contains(pattern_idx);
+}
+
+/// Check if a var reassignment crosses function boundaries
+pub fn isVarReassignmentAcrossFunctionBoundary(self: *const Self, pattern_idx: CIR.Pattern.Idx) bool {
+    if (self.var_function_regions.get(pattern_idx)) |var_function_region| {
+        if (self.getCurrentFunctionRegion()) |current_function_region| {
+            // Manual comparison since Region doesn't have eql method
```

Comment:
I'd actually suggest giving `Region` an `eql` - that way if we ever decode to change how it's represented internally, there will only be one central place to update how its equality logic works.